pkg/cmd/codespace/ssh.go

```diff
@@ -110,25 +110,26 @@ func newSSHCmd(app *App) *cobra.Command {
 }
 
 // SSH opens an ssh session or runs an ssh command in a codespace.
-func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err error) {
+// func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err error) {
+func SSH(app *App, ctx context.Context, sshArgs []string, opts sshOptions) (err error) {
```

Comment:
Why are we moving this method off of `*App` ?

Reply:
In short, because of testing. There's a lot of tight coupling between the "codespaces" and "codespace" packages, which makes mocking of things like "liveshare" pretty challenging. In this case, I moved it off of `*App` so it could be mocked in the tests, but I would love to hear about alternative approaches. I'm not experienced with go testing techniques, and coming from Ruby I really with I could simply `expect(app).to receive(:SSH) { "Yay!" }` 😅 

> If this function needs to be assignable to a func field for testing purposes, it can be easily wrapped in an anonymous func.

Do you mind providing an example of this? I'm not sure how an anonymous function would help here

Reply:
I think it depends on what you are trying to test, from the two tests I see in this PR, none of them require a "spy" like functionality, one wants no errors and the other wants an error both of which can be mocked to behave accordingly.

You've already done the work to create a liveshare session interface, I think what's left is to pass in a "liveshare creator/constructor" into this method and have the Cobra RunE function pass one in. 

ie
```
type liveshareConnector func(context.Context, api.Codespace, *App, bool, string) (Session, error)

func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions, liveshareConnect liveshareConnector) (err error) {
...
session, err := liveshareConnect(ctx, codespace, a, opts.debug, opts.debugFile)
```

This might drop the need to have an `sshClient` field value on the `*App` struct as well, as every external dependency to the method can be mocked or injected.

Reply:
Thanks for this suggestion -- I implemented it and like the results. Now, though, I'm getting stuck at `codespaces.Shell` as it attempts to ssh into the codespace. I'll have to do a little more refactoring to work around that.

Reply:
Give it a `sheller` 😄